docs/proposals/004-initialization

```diff
@@ -0,0 +1,177 @@
+=================
+
+This proposal documents the desired behavior of initialization related language semantics, including defualt constructor, initialization list and variable initialization.
+
+Status
+------
+
+Status: Design Review
+
+Implemtation: N/A
+
+Author: Yong He
+
+Reviewer: 
+
+Background
+----------
+
+Slang has introduced several different syntax around initialization to provide syntactic compatibility with HLSL/C++. As the language evolve, there aree many corners where
+the semantics around initialization are not well-defined, and causing confusion or leading to surprising behaviors.
+
+This proposal attempts to provide a design on where we want to language to be in turns of how initialization is handled in all different places.
+
+Related Work
+------------
+
+C++ has many different ways and syntax to initialize an object: through explicit constructor calls, initialization list, or implicitly in a member/variable declaration.
+A variable in C++ can also be in an uninitialized state after its declaration. HLSL inherits most of these behvior from C++ by allowing variables to be uninitialized.
+
+On the other hand, languages like C# and Swift has a set of well defined rules to ensure every variable is initialized after its declaration.
+
+Proposed Approach
+-----------------
+
+The high level direction for Slang is it needs to help ensuring all decls are properly initialized at its declaration site, while still providing a reasonable level
+of backward compatibility with existing Slang/HLSL code. In this section, we document all concepts and rules to achieve this goal.
+
+## `IDefaultInitializable` interface
+
+The builtin `IDefaultInitializable` interface is defined as:
+```csharp
+interface IDefaultInitializable
+{
+    __init();
+}
+```
+
+Any type that conforms to `IDefaultInitializable` is treated as having the *default-initializable* property in the rest of the discussions. By default, all builtin
+types, such as `int`, `float[]`, `float4` are default-initializable. The values for builtin numeric types after default-initialization are 0.
+The default value for pointers is `nullptr`, and default value for `Optional<T>` is `none`.
+
+## Default-Initializable Type
+
+A type X is default initializable if:
+- It explicitly declares that `X` implements `IDefaultInitializable`.
+- It explicitly provides a default constructor `X::__init()` that takes no arguments, in which case we treat the type as implementing `IDefaultInitializable` even if
+  this conformance isn't explicitly declared.
+- It is a struct type where all its members are default-initializable. A member is considered default-initializable if the type of the member is default-initializable,
+  or if the member has an initialization expression that defines its default value.
+- It is a sized-array type where the element type is default-initializable.
+- It is a tuple type where all element types are default-initializable.
+- It is an `Optional<T>` type for any `T`.
+
```

Comment:
How about the type defined in shader resource type, e.g.:

cbuffer Uniforms
{
	default-initializable-type a;
}

is `a` still considered default-initializable? I think it is.

Reply:
it is considered default-initializable, although it will have no meaning since the location is in a read-only position.